presto-main/src/main/java/com/facebook/presto/execution/scheduler/SqlQueryScheduler.java

```diff
@@ -867,35 +866,47 @@ else if (!result.getBlocked().isDone()) {
 
     private boolean isReadyForExecution(StreamingPlanSection section)
     {
-        SqlStageExecution stageExecution = getStageExecution(section.getPlan().getFragment().getId());
-        if (stageExecution.getState() != PLANNED) {
+        Optional<SqlStageExecution> stageExecution = getStageExecution(section.getPlan().getFragment().getId());
+        if (stageExecution.isPresent() && stageExecution.get().getState() != PLANNED) {
             // already scheduled
             return false;
         }
         for (StreamingPlanSection child : section.getChildren()) {
-            SqlStageExecution rootStageExecution = getStageExecution(child.getPlan().getFragment().getId());
-            if (rootStageExecution.getState() != FINISHED) {
+            Optional<SqlStageExecution> rootStageExecution = getStageExecution(child.getPlan().getFragment().getId());
+            if (!rootStageExecution.isPresent() || rootStageExecution.get().getState() != FINISHED) {
                 return false;
             }
         }
         return true;
     }
 
-    private List<List<StageExecutionAndScheduler>> getStageExecutions(List<StreamingPlanSection> sections)
+    private List<List<StageExecutionAndScheduler>> getOrCreateStageExecutions(List<StreamingPlanSection> sections)
     {
-        return sections.stream()
-                .map(section -> stream(forTree(StreamingSubPlan::getChildren).depthFirstPreOrder(section.getPlan())).collect(toImmutableList()))
-                .map(plans -> plans.stream()
-                        .map(StreamingSubPlan::getFragment)
-                        .map(PlanFragment::getId)
-                        .map(this::getStageExecutionInfo)
-                        .collect(toImmutableList()))
-                .collect(toImmutableList());
+        ImmutableList.Builder<List<StageExecutionAndScheduler>> stageExecutionAndSchedulers = ImmutableList.builder();
+
+        for (StreamingPlanSection section : sections) {
+            if (!getStageExecution(section.getPlan().getFragment().getId()).isPresent()) {
```

Comment:
Maybe I'm wrong, but since this is lazily generated and the plan is a tree, would this ever be present at this point?

Reply:
If we created the section last time, but not all of the stages were scheduled, then the stageExecution would be present.